### PR TITLE
feat: modernize codebase for Python 3.9+ with pydantic validation

### LINKED ISSUE #139: Modernize codebase
- [x] pydantic
- [x] Types
- [x] Pytest
- [x] Pytorch 2.X compatibilities
- [ ] Others

### PR DESCRIPTION
**Pull request by tcoroller**
- Drop Python 3.8 support (EOL Oct 2024), require Python >=3.9
- Add Python 3.12 classifier
- Add pydantic>=2.0 as required dependency
- Bump torch requirement to >=2.8
- Replace validate_data.py with new pydantic-based validators.py
- Add tools/__init__.py module
- Refactor all metrics and loss modules to use new validators
- Add hypothesis-based property tests in tests/strategies.py
- Add shared test fixtures in tests/conftest.py
- Add tests/test_validators.py for new validation logic
- Update dev/environment.yml and run-unittests.sh
- Update ruff target-version to py39
- Add docs/package_overview.md
- Add .hypothesis/ to .gitignore

### CODE REVIEW

**Comment by tcoroller** [comment]
# PR Review Summary: Modernize Codebase + Migrate to uv

This PR has **two major themes**: (1) codebase modernization (Python 3.9+, Pydantic v2, typing, mypy strict) and (2) migration from conda to uv as the development toolchain. Here's a guided tour for reviewers.

---

## 🔴 High Attention — Review Carefully

### 1. Pydantic v2 Validators (`src/torchsurv/tools/validators.py` — NEW file, replaces `validate_data.py`)

This is the biggest behavioral change. The old `validate_data.py` (deleted) used manual checks; the new `validators.py` uses **Pydantic v2 models** for input validation throughout the library.

- **What to check:** Validation logic, error messages, edge cases (empty tensors, NaNs, shape mismatches)
- **Impact:** Every loss function and metric now runs inputs through Pydantic validators
- **Tests:** `tests/test_validators.py` (new)

### 2. Cox Loss Vectorization (`src/torchsurv/loss/cox.py`)

The Cox partial log-likelihood computation was **vectorized** — replacing Python loops with tensor operations. This is a correctness-critical performance change.

- **What to check:** Efron/Breslow tie handling, strata support, numerical equivalence with previous implementation
- **Tests:** `tests/test_cox.py` (heavily modified — compare against lifelines/scikit-survival baselines)

### 3. Source Code — Type Annotations & `__all__` (all files under `src/torchsurv/`)

Every module got **strict mypy typing** and explicit `__all__` exports. All `__init__.py` files now have public API definitions.

- **What to check:** Return types, `Optional` vs `Union` usage, that `__all__` exports match the intended public API
- **Quick scan OK for:** Most changes are additive annotations on existing signatures

---

## 🟡 Medium Attention — Verify Correctness

### 4. CI Workflows (`.github/workflows/` — all 4 files)

All workflows migrated from **conda → uv**:
- `codeqc.yml`: `setup-miniconda` → `setup-uv` + `uv sync` + `uv run`
- `docs.yml`: Same pattern, adds `apt-get install pandoc` (system dep that conda handled implicitly)
- `build-release.yml`: `pip install build twine` → `uv build` + `uvx twine check`
- `torch-compatibility.yml`: pip → `uv sync --override` for PyTorch version matrix

- **What to check:** That each workflow installs the right dependency groups (`--group dev`, `--extra docs`). The torch-compatibility override pattern (`echo 'torch==X.Y.Z' > /tmp/torch-override.txt`) is new.

### 5. `pyproject.toml`

Major changes:
- `requires-python` raised from `>=3.8` to `>=3.9`
- `pydantic>=2.0` added as a **runtime dependency**
- `[dependency-groups] dev` expanded with test deps previously only in `environment.yml` (lifelines, scikit-survival, lightning, loguru, torchvision, etc.)
- `[project.optional-dependencies] docs` expanded (sphinx-book-theme, nbsphinx, sphinxcontrib-bibtex, etc.)
- New `publish` dependency group (build, twine)

- **What to check:** That the dependency list is complete and version bounds are sensible. Pydantic as a runtime dep is a user-facing change.

### 6. Tests (all files under `tests/`)

- New shared fixtures in `conftest.py` and `strategies.py` (Hypothesis strategies)
- Test files updated for new validator API and type annotations
- `test_hypothesis.py` heavily reworked

- **What to check:** That test coverage is maintained. Spot-check a few test files against previous behavior.

---

## 🟢 Low Attention — Quick Scan

### 7. Documentation Updates

- **`docs/devnotes.md`**: uv is now the primary dev setup path; conda moved to collapsed legacy section. All command examples updated.
- **`CONTRIBUTING.md`**: Already was uv-first, removed conda fallback line.
- **`README.md`**: Dev setup section in `<details>` updated to use `uv sync`. End-user install (pip/conda) unchanged.
- **`CLAUDE.md`**: Internal dev guidelines, commands updated to `uv run`.
- **New docs**: `docs/benchmarks.md`, `docs/package_overview.md`, `docs/notebooks/non_medical_applications.ipynb`

### 8. `.pre-commit-config.yaml`

- mypy hook changed from `entry: mypy` to `entry: uv run --no-sync mypy` with explicit `src/` target
- This was needed because mypy is inside `.venv`, not on system PATH

### 9. `dev/environment.yml`

- Added deprecation notice header pointing to `pyproject.toml` + `uv sync`
- File kept for backward compatibility but no longer maintained

### 10. Housekeeping

- `.gitignore` additions
- `CODEOWNERS.md` added
- `communications/` folder added
- `dev/*.sh` scripts: minor cleanups (set -euo pipefail, PYTHONPATH fixes)
- `uv.lock`: New lockfile (auto-generated, ~9k lines — no need to read)

---

## Key Decisions to Validate

| Decision | Rationale |
|---|---|
| Pydantic v2 as **runtime** dependency | Adds ~2MB to install; enables structured validation |
| Drop Python 3.8 support | 3.8 reached EOL Oct 2024 |
| uv as sole dev toolchain | Eliminates dep duplication between pyproject.toml and environment.yml |
| Keep environment.yml (deprecated) | Soft transition for existing conda users |
| `torch>=2.8` minimum | Required for newer torch features used in vectorized cox |

File: docs/notebooks/non_medical_applications.ipynb [hunk 1]
```diff
@@ -0,0 +1,1162 @@
...
+    "    def __len__(self) -> int:\n",
+    "        return len(self.X)\n",
+    "\n",
+    "    def __getitem__(self, idx):\n",
+    "        return self.X[idx], (self.event[idx], self.time[idx])\n",
+    "\n",
+    "\n",
+    "def make_loaders(X, event, time, test_size=0.2, val_size=0.1, batch_size=64):\n",
+    "    \"\"\"Split into train/val/test and return DataLoaders.\"\"\"\n",
+    "    X_tv, X_test, e_tv, e_test, t_tv, t_test = train_test_split(X, event, time, test_size=test_size, random_state=42)\n",
```
Comment (melodiemonod) [inline docs/notebooks/non_medical_applications.ipynb:169]: what does "tv" stand for?

File: docs/notebooks/non_medical_applications.ipynb [hunk 2]
```diff
@@ -0,0 +1,1162 @@
...
+    "    # C-index\n",
+    "    ci = ConcordanceIndex()\n",
+    "    c = ci(log_hz, event_test, time_test)\n",
+    "\n",
+    "    # AUC at median time\n",
+    "    median_t = time_test.median()\n",
+    "    auc_metric = Auc()\n",
+    "    auc_val = auc_metric(estimate=log_hz, event=event_test, time=time_test, new_time=median_t)\n",
+    "\n",
+    "    # Brier Score\n",
```
Comment (melodiemonod) [inline docs/notebooks/non_medical_applications.ipynb:244]: what not add the brier score?

File: docs/notebooks/non_medical_applications.ipynb [hunk 3]
```diff
@@ -0,0 +1,1162 @@
...
+    "    \"\"\"Evaluate a Weibull AFT model. Returns dict of metrics.\"\"\"\n",
+    "    model.eval()\n",
+    "    with torch.no_grad():\n",
+    "        X_test, (event_test, time_test) = next(iter(loader_test))\n",
+    "        params = model(X_test)  # shape (N, 2): [log_scale, log_shape]\n",
+    "        log_scale = params[:, 0:1]\n",
+    "\n",
+    "    # C-index (use negative log_scale as risk score — higher scale = longer survival)\n",
+    "    ci = ConcordanceIndex()\n",
+    "    c = ci(-log_scale, event_test, time_test)\n",
```
Comment (melodiemonod) [inline docs/notebooks/non_medical_applications.ipynb:268]: i disagree with this. we should pass the parameters to get the survival function and use this as risk score. like in the main notebook. happy to do it just let me know if there was a particular other reason to use - log_scale

File: docs/notebooks/non_medical_applications.ipynb [hunk 4]
```diff
@@ -0,0 +1,1162 @@
...
+     "iopub.status.idle": "2026-03-18T18:11:36.178136Z",
+     "shell.execute_reply": "2026-03-18T18:11:36.177857Z"
+    }
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Failure rate: 94.0%  |  Median observed hours: 118\n"
```
Comment (melodiemonod) [inline docs/notebooks/non_medical_applications.ipynb:630]: the failure rate seems very high and the median observed hours super low (less than 5 days of use). I propose to increase the baseline time, like i do in the commit

File: docs/notebooks/non_medical_applications.ipynb [hunk 5]
```diff
@@ -0,0 +1,1162 @@
...
+     "output_type": "stream",
+     "text": [
+      "Event rate: 64.3%  |  Median inter-event gap: 73 days\n"
+     ]
+    }
+   ],
+   "source": [
+    "# ── Synthetic Extreme Weather Data ───────────────────────────────────────────\n",
+    "rng = np.random.default_rng(4)\n",
+    "N = 600\n",
```
Comment (melodiemonod) [inline docs/notebooks/non_medical_applications.ipynb:927]: why N = 600 and not 800 like all other examples?

File: docs/notebooks/non_medical_applications.ipynb [hunk 6]
```diff
@@ -0,0 +1,1162 @@
...
+      "text/plain": [
```
Comment (melodiemonod) [inline docs/notebooks/non_medical_applications.ipynb:970]: why do we make of the validation that is increasing?

Comment (melodiemonod) [reply docs/notebooks/non_medical_applications.ipynb:970]: overfitting?

File: docs/notebooks/non_medical_applications.ipynb [hunk 7]
```diff
@@ -0,0 +1,1162 @@
...
+    "| 1 | **Customer Churn** (Telecom) | Subscription cancellation | Duration as subscriber | Cox |\n",
+    "| 2 | **Employee Attrition** (HR) | Employee resignation | Tenure | Weibull AFT |\n",
+    "| 3 | **Equipment Failure** (Maintenance) | Machine breakdown | Operating hours | Cox |\n",
+    "| 4 | **Loan Default** (Finance) | Default event | Months since origination | Weibull AFT |\n",
+    "| 5 | **Extreme Weather Events** (Meteorology) | Threshold exceedance | Days between events | Cox |\n",
+    "\n",
+    "Each section:\n",
+    "- Generates **synthetic data** with realistic distributions (fixed seeds for reproducibility)\n",
+    "- Points to a **real public dataset** you can use as a drop-in replacement\n",
+    "- Trains both a **Cox** or **Weibull AFT** model using a simple MLP backbone\n",
```
Comment (melodiemonod) [inline docs/notebooks/non_medical_applications.ipynb:27]: this is not true , you only train one of them

File: docs/devnotes.md [hunk 8]
```diff
@@ -11,87 +11,110 @@ pre-commit install
```
Comment (melodiemonod) [inline docs/devnotes.md:11]: i've tried the new steps with uv and it all works

File: src/torchsurv/tools/validators.py [hunk 9]
```diff
@@ -0,0 +1,503 @@
...
+                [5., 6.]])
+    """
+    if log_params.dim() == 1:
+        log_params = log_params.unsqueeze(1)
+    if log_params.dim() == 2 and log_params.size(1) == 1:
+        log_params = torch.hstack((log_params, torch.zeros_like(log_params)))
+    return log_params
+
+
+def validate_time_varying_log_hz(time_sorted: torch.Tensor, log_hz_sorted: torch.Tensor) -> None:
```
Comment (melodiemonod) [inline src/torchsurv/tools/validators.py:472]: I think we can delete this function. It has been replaced by TimeVaryingCoxInputs.

File: src/torchsurv/loss/cox.py [hunk 10]
```diff
@@ -72,49 +74,61 @@ def _partial_likelihood_efron(
...
+        # denominator_naive[k] = sum of exp(log_hz) for risk set R[k] = {i : time[i] >= time_unique[k]}
+        # Since data is sorted, R[k] is a suffix of the sorted array starting at first_idx[k].
+        # reverse_cumsum[i] = sum(exp_hz[i:]), so denominator_naive[k] = reverse_cumsum[first_idx[k]].
+        reverse_cumsum = exp_hz.flip(0).cumsum(0).flip(0)  # (N,)
+        first_idx = _searchsorted(time_sorted, time_unique)  # (K,)
+        denominator_naive = reverse_cumsum[first_idx]
+
+        # denominator_ties[k] = sum of exp(log_hz) for event subjects at time_unique[k]
+        denominator_ties = torch.zeros(K, device=device)
+        denominator_ties.scatter_add_(0, event_idx, exp_hz[event_sorted])
```
Comment (melodiemonod) [inline src/torchsurv/loss/cox.py:117]: I believe the code was lighter and clearer before. in the if else statements. Why the change?

File: CODEOWNERS.md [hunk 11]
```diff
@@ -14,3 +14,8 @@
 * David Ohlssen (**Novartis**)
 * Berkman Sahiner (**FDA**)
 * Nicholas Petrick (**FDA**)
+
+
+### Contributors
+
+* Yao Chen (**Novartis**)
```
Comment (melodiemonod) [inline CODEOWNERS.md:21]: should we make the changes with the acknowledgments as we discussed?

Comment (tcoroller) [reply CODEOWNERS.md:18]: Fixed

File: src/torchsurv/loss/cox.py (hunk 10 above)
Comment (tcoroller) [reply src/torchsurv/loss/cox.py:117]: vectorized version, but I can revert this part for readability.

File: src/torchsurv/tools/validators.py (hunk 9 above)
Comment (tcoroller) [reply src/torchsurv/tools/validators.py]: It was to support deprecated features, but good point - removing it

File: docs/notebooks/non_medical_applications.ipynb (hunk 1 above)
Comment (tcoroller) [reply docs/notebooks/non_medical_applications.ipynb:169]: TrainValidation. They are combined first, then split

File: docs/notebooks/non_medical_applications.ipynb (hunk 3 above)
Comment (tcoroller) [reply docs/notebooks/non_medical_applications.ipynb:268]: @melodiemonod  can you directly change this?

File: docs/notebooks/non_medical_applications.ipynb (hunk 2 above)
Comment (tcoroller) [reply docs/notebooks/non_medical_applications.ipynb:244]: @melodiemonod same here
